[NORI] Jsonparser step1

NORIKIM opened the pull request:
JSON step1 단순 List 분석 PR 합니다.

NORIKIM committed:
사용자의 입력을 확인하고(CheckInput) 형식이 맞지 않으면 오류 메시지를 출력(ErrorMessage)

NORIKIM committed:
사용자의 입력을 받도록 문구를 출력하고(readInput) 입력값을 콤마로 분리하여 배열을 생성(splitInput)

NORIKIM committed:
사용자의 입력값을 부울,문자열,숫자로 분리해서 가져옴

NORIKIM committed:
총 입력 데이터 개수를 파악하고 입력값에서 각 타입별로 개수를 출력하도록 실행

NORIKIM committed:
총 데이터의 개수와 각 타입을 구분하여 출력시킴

NORIKIM committed:
사용자의 입력값에서 각 타입별로 분리하기위한 정규표현식 함수 구현

NORIKIM committed:
JSONParser 코드

NORIKIM committed:
JSONParser 테스트 코드

godrm commented on JSONParser/JSONParser/JSONData.swift:
```diff
@@ -0,0 +1,37 @@
+        self.typeBoolean = typeBoolean
+        self.typeString = typeString
+        self.typeNumber = typeNumber
+    }
```
항상 생성자에서 이렇게 초기화한 내부 var 변수들은 private 으로 만들어서 외부에서 변경불가능하도록 만드세요.

godrm commented on JSONParser/JSONParser/JSONData.swift:
```diff
@@ -0,0 +1,37 @@
+        let typeString = RegularExpression.bringData(from: splitInput, regex: "\"{1}+[A-Z0-9a-z]+\"{1}")
+        let typeNumber = splitInput.count - (typeBoolean.count + typeString.count)
+        
+        let data = JSONData.init(typeBoolean: typeBoolean.count, typeString: typeString.count, typeNumber: typeNumber)
```
오 정규표현식을 벌써 사용했군요!
근데 JSONData 는 배열을 갖고 있지 않네요? 
data 변수는 출력을 위한 카운트만 갖고 있는데, 카운트 값은 부가적인 결과입니다.
이번 미션에서 집중해야 하는 것은 문자열로 만들어진 JSON을 스위프트 Array에 타입별로 어떻게 넣어놓을것인가 생각해야 합니다. 
`[ "JK", 33, false ]` JSON 데이터가 그대로 `"JK"`, `33`, `false` 값들을 포함하는 배열이 있어야 합니다. 그 배열 내부에 요소가 얼마나 있는지는 출력할 때만 필요한 거구요

NORIKIM committed:
var 변수를 private으로 수정

NORIKIM committed:
여러 타입을 받아 올 수 있도록 JsonValue 프로토콜 생성

NORIKIM committed:
메소드 설명 주석 추가

NORIKIM committed:
불필요한 메소드 bringData 삭제

NORIKIM committed:
사용자이 입력에 대괄호가 있는지 확인하는 함수 추가, 사용자의 입력을 검사하는 총괄 함수 추가

NORIKIM committed:
괄호가 포함되는지 확인하는 작업 수정

NORIKIM committed:
프로토콜을 JsonValue를 이용하여 다양한 타입을 저장하도록 JSONData생성

NORIKIM committed:
주석 입력

NORIKIM committed:
정규표현식

1.  정규표현식 상수 생성
2. 정규표현식을 사용하여 문자열 검사 시 확인 결과 리턴 메소드 생성(containsMatch)
3. 배열에서 문자열을 확인하여 숫자,문자열,부울 형태를 리턴하는 메소드 생성(intCasting,stringCasting,boolCasting)
4. JSONData에 값들을 넣을 수 있는 메소드 생성(makeJsonData)

NORIKIM committed:
입력값 확인

1. 입력이 정확하지 않을 경우 출력 메시지 수정
2. 입력에 공백을 확인하는 메소드 추가(hasWhiteSpace)
3. 입력에 대괄호를 확인하는 메소드 추가(hasParentheses)
4. 입력이 유효한지 총괄 메소드 추가(validInput)

NORIKIM committed:
수정사항 없음

NORIKIM committed:
입력값에 대한 확인 및 출력을 하는 메인

NORIKIM committed:
입력값을 딕셔너리로 정리 및 확인하여 데이터 내용을 출력

1. 딕셔너리를 만드는 메소드 생성(countData)
2. 데이터의 개수에 따라 내용을 출력하는 메소드 생성(printData)
3. 모든 데이터의 내용을 출력하는 메소드 생성(printDatas)

NORIKIM committed:
프로토콜 JsonValue에 들어가는 값들의 타입을 확인하도록 확장

NORIKIM committed:
메소드 intCasting을 private으로 변경

NORIKIM committed:
JSONParser 코드

NORIKIM committed:
JSONParser 테스트 코드

NORIKIM commented:
드디어 PR 보냅니다...ㅠ
근데 문자열 숫자에 대한 처리가 조금 미흡해서 어떻게 해결해야 할지 더 고민해봐야 할 것 같습니다.

godrm commented on JSONParser/JSONParser/OutputView.swift:
```diff
@@ -1,43 +1,61 @@
+            }
+        }
+        return data
+    }
```
countData() 이 부분은 다음 단계를 하면서 별도 객체로 분리해보세요.

godrm commented:
1단계에서 해야할 부분보다 더 많이 정규표현식까지 써서 잘 구현했네요.
다음 단계를 진행해보세요.
